garglk/config.cpp

```diff
@@ -38,6 +38,10 @@
 bool gli_utf8input = true;
 bool gli_utf8output = true;
 
+int gli_hires = 1;
+float gli_backingscalefactor = 1.0f;
```

Comment:
I didn't write this code and I don't pretend to understand it, but it seems that the gli_backingscalefactor variable is calculated at lines 368-374 in sysmac.mm, in the wininit method. It is a way to convert size in points to actual pixels. I think it will always be 2 on a Retina screen and 1 on a low resolution screen.

Reply:
Common pixel ratios include 1, 1.5, 2, 3, but can also be 1.33, 1.67, 2.46, 2.61, 2.75... though those probably only occur on phones. Windows will let you enter any number (as a percentage) between 100 and 500, so someone might be running Gargoyle on something totally weird.

Reply:
Aside: I raised the idea of Glk functions to expose the pixel ratio here: https://github.com/erkyrath/glk-dev/issues/3

Reply:
@angstsmurf Somehow I missed that assignment in sysmac.mm; thanks.

@curiousdannii I know nothing about hi res displays, so I'll be asking questions from ignorance here; would the Windows pixel ratio be used to calculate the backing scale on Windows?  Do you have a hidpi display on Windows, and if so, how does Gargoyle look on it?

If it's easy to get Windows hidpi support then it's worth putting it in, but unfortunately, I don't think I'd have a way to properly test it.

Reply:
It just looks blurred, as it's just upscaling from 100%.

You should be able to test it by changing one of your monitor's scales in the display settings. You don't need an actual hidpi screen.

Reply:
The gli_backingscalefactor for Windows  was computed in syswin.c:

https://github.com/garglk/garglk/commit/0e0aeeef6c331f298bb44838e8ac972f9642db6a#diff-dcdea0104f827a9d6043e4e1066cde7a21d77ead3c2d972dec57dcea055d1120L282

HiDPI support makes a big difference visually. I'd like to invest a few hours to see if Qt+HiDPI is possible on Windows an Linux. But maybe it's just me who is picky about the blurry fonts ;)

@angstsmurf : do you have a list of games with Hi-DPI issues apart from Andromeda 1983, for testing purposes?

Reply:
> You should be able to test it by changing one of your monitor's scales in the display settings. You don't need an actual hidpi screen.

I only have Windows in a VM so I'm not sure how well that'll work, but I can give it a go.

> HiDPI support makes a big difference visually. I'd like to invest a few hours to see if Qt+HiDPI is possible on Windows an Linux. But maybe it's just me who is picky about the blurry fonts ;)

Thanks for giving this a look!

Reply:
@hkleinsorgen A good test is probably [graphwintest.gblorb](https://www.eblong.com/zarf/glulx/graphwintest.gblorb), which is specifically written to test graphics windows, which seems to be where the bugs are.

Andromeda 1983 uses the "Location Images" extension, and all games which uses this seem to have the same problem. Another example is [Wolfsmoon](https://ifdb.org/viewgame?id=npxjxi6n9oak7zo7).

Reply:
I'm facing some difficulties building garglk for Windows, neither cross compiling nor MSYS2 worked for me so far. So I suggest that you go ahead with macOS only for the time being.

Reply:
@hkleinsorgen The Windows build is, to put it mildly, idiosyncratic vis-à-vis normal Windows development.  It expects a Unix-like environment with system packages that can be found via CMake, and probably pkg-config as well.

When I have some time I can see about making Windows support easier to handle, but I'm not sure I want to go back to bundling libraries as Gargoyle used to do.  Using existing packages is much cleaner, but if there are ways to be more inclusive (e.g. working with MSYS2, something I'm unfamiliar with) I'm all for it.

If you have an interest in a working Windows build, and don't mind a bit of setup work, you can use Gargoyle's MinGW Github Action as a guide.  It makes use of a prebuilt MinGW environment.  Setup is involved, but once it's done, it should be smooth.  Rough steps:

1. Install an Ubuntu 20.04 VM (this matches what GH Actions provide)
2. Install the following packages: build-essential cmake g++-mingw-w64 gcc-mingw-w64 mingw-w64-tools wget
3. Run `update-alternatives --set i686-w64-mingw32-gcc /usr/bin/i686-w64-mingw32-gcc-posix` and `update-alternatives --set i686-w64-mingw32-g++ /usr/bin/i686-w64-mingw32-g++-posix`
4. `wget https://github.com/garglk/assets/raw/master/mingw/mingw-i686.tar.xz && tar xfv mingw-i686.tar.xz -C /`
5. Run `NO_INSTALLER=1 ./windows.sh` from the top level of the `garglk` repository

At this point, if all goes well, you have `build-mingw` as the CMake build directory, and `build/dist` as the install target.  The script _should_ copy all necessary DLLs over, so from now on, all you have to do is run `make install` in the `build-mingw` directory to be able to quickly test the new changes, directly in `build/dist`.

I realize this is completely annoying, but once steps 1-4 are done, Windows development should go pretty smoothly.